### PR TITLE
New variant of mesh_loop.

### PR DESCRIPTION
**Pull request by luca-heltai**
This variant uses a class and its member functions, providing another interface similar to WorkStream::run

### CODE REVIEW

File: doc/news/changes/minor/20190323LucaHeltai [hunk 1]
```diff
@@ -0,0 +1,3 @@
+New: Added a variant of mesh_loop that takes a class and its member functions as workers and copiers.
```
Comment (jppelteret) [inline doc/news/changes/minor/20190323LucaHeltai]: `mesh_loop` -> `mesh_loop()`

File: include/deal.II/meshworker/mesh_loop.h [hunk 2]
```diff
@@ -392,6 +392,116 @@ namespace MeshWorker
                     queue_length,
                     chunk_size);
   }
+
+  /**
+   * This is a variant of the mesh_loop function, that can be used for worker
```
Comment (jppelteret) [inline include/deal.II/meshworker/mesh_loop.h]: `mesh_loop()`

File: include/deal.II/meshworker/mesh_loop.h [hunk 3]
```diff
@@ -392,6 +392,116 @@ namespace MeshWorker
...
+   * and copier functions that are member functions of a class.
+   *
+   * The argument passed as @p end must be convertible to the same type as @p
+   * begin, but doesn't have to be of the same type itself. This allows to
+   * write code like <code>mesh_loop(dof_handler.begin_active(),
+   * dof_handler.end(), ...</code> where the first is of type
```
Comment (jppelteret) [inline include/deal.II/meshworker/mesh_loop.h]: `dof_handler.end(), ...)</code>`

(missing end parenthesis)

File: include/deal.II/meshworker/mesh_loop.h [hunk 4]
```diff
@@ -392,6 +392,116 @@ namespace MeshWorker
...
+   * DoFHandler::active_cell_iterator whereas the second is of type
+   * DoFHandler::raw_cell_iterator.
+   *
+   * The @p queue_length argument indicates the number of items that can be
+   * live at any given time. Each item consists of @p chunk_size elements of
+   * the input stream that will be worked on by the worker and copier
+   * functions one after the other on the same thread.
+   *
```
Comment (jppelteret) [inline include/deal.II/meshworker/mesh_loop.h]: I think that our less experienced users sometimes struggle with the syntax of these sorts of functions (I know that I certainly did when first learning to use the workstream functions). Could we put an small example showing how to use this? From your test:

```
template <int dim, int spacedim>
class TestClass
{
public:
  void
  cell_worker(const CellIteratorType &cell, ScratchData &, CopyData &);

  ...
};

  ...
  TestClass<dim, spacedim> test_class;
  ScratchData              scratch;
  CopyData                 copy;

  mesh_loop(tria.begin_active(),
            tria.end(),
            test_class,
            &TestClass<dim, spacedim>::cell_worker,
            &TestClass<dim, spacedim>::copier,
            scratch,
            copy,
            assemble_own_cells);
```

File: tests/meshworker/mesh_loop_class_01.cc [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+class TestClass
+{
+public:
+  using CellIteratorType =
+    typename Triangulation<dim, spacedim>::active_cell_iterator;
+
+  void
+  cell_worker(const CellIteratorType &cell, ScratchData &, CopyData &)
+  {
+    deallog << "Workgin on cell " << cell << std::endl;
```
Comment (jppelteret) [inline tests/meshworker/mesh_loop_class_01.cc]: Workgin -> Working

File: include/deal.II/meshworker/mesh_loop.h (hunk 4 above)
Comment (jppelteret) [reply include/deal.II/meshworker/mesh_loop.h]: Erg, I'm sorry, somehow I managed to use a stale version of this code example in my comment. I've fixed this with what I had intended to suggest to you, and have added an extra commit to (hopefully) fix the tester error.

**Comment by luca-heltai** [comment]
Thanks JP!

**Pull request merged**